### ISSUE TITLE
Large discrepancy between predicted camera intrinsics and ground truth

### ISSUE THREAD

**Issue by Lyuyarn** [facebookresearch/map-anything#38]
Thank you for your excellent work and for sharing this powerful 3D reconstruction tool with the community. 

However, I have encountered an issue regarding the accuracy of the predicted camera intrinsics during inference. The predicted values differ significantly from the ground truth parameters.

Inference code:
import os
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"

import torch
from mapanything.models import MapAnything
from mapanything.utils.image import load_images

device = "cuda" if torch.cuda.is_available() else "cpu"
model = MapAnything.from_pretrained("facebook/map-anything", local_files_only=True).to(device)

images = "./input/cola"
views = load_images(images)

predictions = model.infer(
    views,
    memory_efficient_inference=False,
    use_amp=True,
    amp_dtype="bf16",
    apply_mask=True,
    mask_edges=True,
    apply_confidence_mask=False,
    confidence_percentile=10,
)

for i, pred in enumerate(predictions):
    intrinsics = pred["intrinsics"]
    print(intrinsics)

Predicted intrinsics:

<img width="775" height="159" alt="Image" src="https://github.com/user-attachments/assets/2a760dca-d8be-48aa-8cc4-398bd42f2fc1" />

Ground truth intrinsics:

<img width="1188" height="165" alt="Image" src="https://github.com/user-attachments/assets/322e37e7-fd30-4d0b-b593-2351219f4131" />

**Comment by Lyuyarn** [comment]
Furthermore, I have tried multi-modal inference by explicitly providing the camera intrinsics as input for each view. However, even when supplying the ground truth intrinsics, the predicted intrinsics output by the model still show a large discrepancy compared to the actual values.

Thank you again for your support and clarification!

inference code:

[multi_modal_infer.py](https://github.com/user-attachments/files/22580543/multi_modal_infer.py)

output:

<img width="865" height="565" alt="Image" src="https://github.com/user-attachments/assets/d5470cd6-df33-4db4-89d1-1d1849010a40" />

**Comment by Lyuyarn** [comment]
By the way, I encountered another issue during multi-modal inference. In my previous comment, I only provided camera intrinsics, but then I added depth_z as well. During inference, I got a runtime error related to tensor shape mismatch:
```python
pts3d_cam = depth_z * ray_directions_unit_plane
```
To resolve this, I modified line 249 in mapanything/utils/inference.py to:
```python
pts3d_cam = depth_z.unsqueeze(-1) * ray_directions_unit_plane
```
I'm not sure if this is the correct fix, but it allowed the inference to proceed and produced new results.

output:

<img width="860" height="517" alt="Image" src="https://github.com/user-attachments/assets/2931a259-e0be-4f59-9093-3414deed6804" />

However, the results still do not seem to be correct.

**Comment by luisCartoGeo** [comment]
Sorry to intrude, but you are considering that Map Anything resamples the original image to a lower resolution, which causes the intrinsic parameters to differ.

**Comment by Lyuyarn** [comment]
> Sorry to intrude, but you are considering that Map Anything resamples the original image to a lower resolution, which causes the intrinsic parameters to differ.

Thank you for your helpful reminder!

Based on your suggestion, I adjusted the resize mode in mapanything/utils/image.py at line 335. I set resize_mode="fixed_size" and size=(1920, 1080) to match the original image resolution. However, due to model constraints, the resolution must be divisible by 14, so the final image size provided to the model is (1918, 1078).

Despite this adjustment, the predicted camera intrinsics fx still differ from the actual parameters by around 200. 

Is there anything else I should try adjusting? Thanks again for your support!

<img width="928" height="276" alt="Image" src="https://github.com/user-attachments/assets/19ca695e-0205-475b-a653-c8b597855fc3" />

**Comment by SirWyver** [comment]
Since MapAnything was trained on lower resolutions that your input image, I'd recommend to rather scale the predicted intrinsics back back to your original resolution using the image sized used during inference.
According to the results you shared it should be something like
```
pred_fx = 325.5880 [at what I assume was width=518] -> 325.5880 * 1920/518 = **1206.81**
pred_fy = 326.9598 [at what I assume was height=294] -> 326.9598 * 1080/294 = **1201.076**
```
Please verify that those are the correct image resolution used during inference.

**Comment by Nik-V9** [comment]
Hi, Thanks for trying out MapAnything! As @luisCartoGeo and @SirWyver mentioned, the main issue is that the infer call resizes and crops the images during inference.

The main reason why we do this resizing and cropping to a fixed mapping (with longest side being 518) is because we observe that the performance of the model degrades significantly on resolutions it is not trained on (particularly the camera calibration/ray prediction).

So as suggested by the others, I would recommend rescaling and adjusting the outputs to match the original resolution (this might be a bit tricky if the original image is not a standard aspect ratio and there is some cropping involved - for this case, I recommend forcing a manual resize to the closest aspect ratio with 518 as longest side).
